wom/src/main/scala/wdl4s/wdl/WdlCall.scala

```diff
@@ -56,28 +55,37 @@ object WdlCall {
 
   private def buildWomNodeAndInputs(wdlCall: WdlCall): ErrorOr[CallWithInputs] = {
 
-    val inputToOutputPortValidations: Iterable[ErrorOr[(String, GraphNodePort.OutputPort)]] = for {
-      (inputName, expr) <- wdlCall.inputMappings
-      variable <- expr.variableReferences
-      parent <- wdlCall.parent
-      node <- parent.resolveVariable(variable.terminal.sourceString)
-      outputPort = outputPortFromNode(node, variable.terminalSubIdentifier)
-      inputNameToOutputPortMapping = outputPort map { op => inputName -> op }
-    } yield inputNameToOutputPortMapping
+    def resolveVariable(v: AstTools.VariableReference): ErrorOr[(String, GraphNodePort.OutputPort)] = {
+      val name =  v.fullVariableReferenceString
+      for {
+        parent <- wdlCall.parent.toValidNel("Call has no parent?")
+        node <- parent.resolveVariable(v.terminal.sourceString).toValidNel(s"Failed to resolve variable $name")
+        outputPort <- outputPortFromNode(node, v.terminalSubIdentifier)
+      } yield name -> outputPort
+    }
 
-    val inputToOutputPortValidation: ErrorOr[List[(String, GraphNodePort.OutputPort)]] = inputToOutputPortValidations.toList.sequence
 
-    // TODO: When WomExpressions wrap WdlExpressions, we can switch this to use expressionBasedInputs instead of portBasedInputs:
-    val combinedValidation = (inputToOutputPortValidation |@| wdlCall.callable.womDefinition) map { (_, _) }
+    val graphNodeInputExpressionValidation: ErrorOr[List[GraphNodeInputExpression]] = {
+      def graphNodeInputExpression(tuple: (String, WdlExpression)): ErrorOr[GraphNodeInputExpression] = {
+        val (name, wdlExpression) = tuple
+        for {
+          resolvedVariables <- wdlExpression.variableReferences.toList traverse resolveVariable
+          womExpression = WdlWomExpression(wdlExpression, None)
+        } yield GraphNodeInputExpression(name, womExpression, resolvedVariables.toMap)
+      }
+
+      wdlCall.inputMappings.toList traverse graphNodeInputExpression
+    }
+
+    val combinedValidation = (graphNodeInputExpressionValidation |@| wdlCall.callable.womDefinition).tupled
 
     for {
       combined <- combinedValidation
       (inputToOutputPort, callable) = combined
```

Comment:
ToL (I know this is old code) I believe you can directly assign tuples in for comprehension, i.e. combine this line and above.